### ISSUE TITLE
Global Config File

### ISSUE THREAD

**Issue by dbwiddis** [oshi/oshi#888]
Create a global configuration file with the following specifications:
 - store the file in `src/main/resources`
 - easily read the file into a Java `Properties` object
 - have a method in the `SystemInfo` class to get the current `Properties` object
   - <s>Object is mutable, can get/set properties on the object.</s>
 - as an initial implementation
   - set a global default WMI timeout property and make it the default in the `WmiUtil` class.
     - If property not set or negative, defaults to infinite
   - set a global path for `/proc` filesystem and make it a constant in the `ProcUtil` class.
     - If property not set, defaults to `/proc`.

**Comment by cilki** [comment]
The method that yields the `Properties` object needs to be `static` so users can modify the default configuration before the `SystemInfo` is created. OSHI classes that use configuration values should consume them on initialization rather than repeatedly querying the `Properties` object. This prevents the user from dynamically updating the values (which I think is ideal).

Also, naming conventions like `oshi.example.value` should be used throughout the config.

Otherwise this sounds good and I can make a PR when I have time.

**Comment by dbwiddis** [comment]
One of the paradigms I was trying to get at with 4.x is sort of a "poor man's thread safety" where if you create a new `SystemInfo` object, everything you get from that is a unique object.  Static properties would end up being shared across threads and require jumping through some more hoops to be thread safe, unless we made the object immutable.

But what if I want to change these global defaults (e.g., WMI default timeout) at runtime?  Recall some of our users will be using OSHI in long-running systems that don't reboot the JVM.

**Comment by dbwiddis** [comment]
The `.property` extension is redundant. The file itself will be `oshi.default.properties`.  I do like a dotted convention that corresponds to the class name and attribute/getter name where it's read on initialization, though, e.g., `oshi.util.platform.os.Windows.WmiTimeout = 500`.

**Comment by dbwiddis** [comment]
Another (probably better) option to SystemInfo is a separate `OshiProperties` class.

**Comment by cilki** [comment]
> The .property extension is redundant

I should have given a real example. I mean that each property name in the config will be in dotted notation exactly like you said. So by `oshi.example.property` I meant something like `oshi.wmi.timeout`. Sorry for the confusion.

**Comment by cilki** [comment]
> But what if I want to change these global defaults (e.g., WMI default timeout) at runtime?

If the `Properties` is static, then you would need to obtain a new `SystemInfo` after making changes to notice the updated values. Now I'm realizing that it might be useful to have multiple `SystemInfo`s with different configurations, so I think you're right that each `SystemInfo` needs its own non-static configuration.

It might be nice to have a static `Properties` somewhere that each `SystemInfo` inherits from so the defaults are read from disk only once. Optionally we could allow the user to modify this `Properties` so each `SystemInfo` is automatically created with their modified defaults.

**Comment by dbwiddis** [comment]
So I've been thinking about this for a while, and researching "best practices" online, most of which oppose "global constants" or configs and think all static variables are evil.  

At the moment I'm leaning towards a static method that reads a config file and stores it in a (read only) (static) Properties object.  For "runtime configuration" each class should/will have overrides. (e.g., any individual WMI query can set its own timeout).    This still leaves a bit of a "middle ground" in terms of the "we never shut down the JVM" users, and I wouldn't be opposed to a method to "reload" the config file.

But I think changing the "global defaults" is probably not something to work into the design.   A config file of "probably will never change" values, supplemented by individual class overrides, is likely the way to go.

**Comment by spyhunter99** [comment]
I guess it depends on the use case. I can see arguments for several different patterns. I've used singleton patterns for configuration objects with reasonable success. I've also used what you have described with good success too. 

If you are open to a new dependency, commons-config has a number of interesting uses, such as properties file or xml file, automatic reloads when the config changes, settings can also be captured via system properties or env variables.

Going down the rabbit hole of use cases, i can see someone building a ui around oshi that may have some user configurable settings, such as timeouts etc. So i think updating config settings at runtime should be a requirement. My specific use case is headless so once i'm past the initial startup, i'm not messing around too much with config settings. Setting the wmi timeout is vital as i've had issues with wmi never responding under cases of high system load. This causes thread leaks and eventual death. For me, long term stability and accuracy of data is important

**Comment by dbwiddis** [comment]
Yeah, I mentioned commons-config in #208 when I first thought about this, but I want to avoid another dependency.  And really, the more I think about it, the less I think it's a good idea.   Googling for "best practices for java config file for global constants" generally results in "don't do it" answers.

**Comment by dbwiddis** [comment]
So I'm back to thinking a "read file once on startup and set constants" solution might work.  It won't be an issue with multithreading because it's only read once.  The discussion in #893 about a custom `/proc` path would be another good candidate here.

**Comment by cilki** [comment]
So to summarize the plan:

There will be an immutable `Properties` object of global defaults, loaded only once from the classpath, that each `SystemInfo` instance will copy upon initialization to become a local configuration. The user can change the local configuration and updates will be noticed by the util classes.

Does that sound right? I think it will work as long as each util class can have a reference to its parent `SystemInfo`.

**Comment by dbwiddis** [comment]
> There will be an immutable `Properties` object of global defaults, loaded only once from the classpath, that each `SystemInfo` instance will copy upon initialization to become a local configuration. 

True except for "each".  It will be a static object, accessible by direct call to the class (`SystemInfo.getConfig()`).

The use case for this is something generally well-known at initialization that won't be changed.   The (usually Util) classes which currently have private static constants in them will initialize those constants by a call to the SystemInfo class.

**Comment by dbwiddis** [comment]
Continuing to think of this... we may not want to put anything more on the "SystemInfo" object.  So far our use cases apply around "Util" classes (ProcUtil, WMI) which we're also talking about being "drivers".   We could have an entirely separate PropertiesUtil class with appropriate methods to update/load in new properties (setUtilProperties,  reloadUtilPropertiesFromFile) and getter methods for individual properties, e.g., getIntProperty("WmiDefaultTimeout").

**Comment by cilki** [comment]
Storing the configuration statically in `SystemInfo` or in a `PropertiesUtil` class also has the effect that there's no such thing as local configuration (all `SystemInfo`s have the same configuration values). Is this something that we should support? Users might want to maintain multiple `SystemInfo` "sessions" with conflicting configurations.

If the util classes are going to become non-static "drivers" with a parent `SystemInfo`, then having a separate configuration for each `SystemInfo` would be easy.

Regardless, I'll build a simple `PropertiesUtil` static class to get this off the ground.

**Comment by dbwiddis** [comment]
> there's no such thing as local configuration (all SystemInfos have the same configuration values). Is this something that we should support? 

In my opinion, yes.  That will clearly define the scope of this config object.  I do not envision these settings being a runtime user preference.  There are essentially three scopes of config-type settings:
 - Global constants that never change (but are loaded from a config file to allow customizable "consttants"
 - Config settings that are sticky and apply globally (e.g., WMI timeout for all queries)
 - Config settings for a particular operation (e.g., WMI timeout for the Sensors query)

Something like the /proc alternative location shouldn't change during runtime, so I see that in the higher level config.  (In the rare situation you don't want to restart the JVM/OSHI after installing a new proc-equivalent, reloading the config can handle that.)

**Comment by dbwiddis** [comment]
> f the util classes are going to become non-static "drivers" with a parent SystemInfo, then having a separate configuration for each SystemInfo would be easy.

I'm open to this as well, but this means passing the Properties from SystemInfo down the tree to every other object so they all have the reference.   This would work generally into my thought of a "thread safe as long as you create a new SystemInfo instance" middle-ground.   It's just a lot more overhead.

I'll leave the choice to professional software developers. :)

**Comment by dbwiddis** [comment]
Done in #896
